tablizer.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "TABLIZER 1"
.TH TABLIZER 1 "2025-10-13" "1" "User Commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tablizer \- Manipulate tabular output of other programs
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 2
\&    Usage:
\&      tablizer [regex,...] [\-r file] [flags]
\&    
\&    Operational Flags:
\&      \-c, \-\-columns string               Only show the speficied columns (separated by ,)
\&      \-v, \-\-invert\-match                 select non\-matching rows
\&      \-n, \-\-numbering                    Enable header numbering
\&      \-N, \-\-no\-color                     Disable pattern highlighting
\&      \-H, \-\-no\-headers                   Disable headers display
\&      \-s, \-\-separator <string>           Custom field separator (maybe char, string or :class:)
\&      \-k, \-\-sort\-by <int|name>           Sort by column (default: 1)
\&      \-z, \-\-fuzzy                        Use fuzzy search [experimental]
\&      \-F, \-\-filter <field[!]=reg>        Filter given field with regex, can be used multiple times
\&      \-T, \-\-transpose\-columns string     Transpose the speficied columns (separated by ,)
\&      \-R, \-\-regex\-transposer </from/to/> Apply /search/replace/ regexp to fields given in \-T
\&      \-j, \-\-json                         Read JSON input (must be array of hashes)
\&      \-I, \-\-interactive                  Interactively filter and select rows
\&      \-g, \-\-auto\-headers                 Generate headers if there are none present in input
\&      \-x, \-\-custom\-headers a,b,...       Use custom headers, separated by comma
\&
\&    Output Flags (mutually exclusive):
\&      \-X, \-\-extended                     Enable extended output
\&      \-M, \-\-markdown                     Enable markdown table output
\&      \-O, \-\-orgtbl                       Enable org\-mode table output
\&      \-S, \-\-shell                        Enable shell evaluable output
\&      \-Y, \-\-yaml                         Enable yaml output
\&      \-J, \-\-jsonout                      Enable JSON output
\&      \-C, \-\-csv                          Enable CSV output
\&      \-A, \-\-ascii                        Default output mode, ascii tabular
\&      \-L, \-\-hightlight\-lines             Use alternating background colors for tables
\&      \-o, \-\-ofs <char>                   Output field separator, used by \-A and \-C. 
\&      \-y, \-\-yank\-columns                 Yank specified columns (separated by ,) to clipboard,
\&                                         space separated
\&
\&    Sort Mode Flags (mutually exclusive):
\&      \-a, \-\-sort\-age                     sort according to age (duration) string
\&      \-D, \-\-sort\-desc                    Sort in descending order (default: ascending)
\&      \-i, \-\-sort\-numeric                 sort according to string numerical value
\&      \-t, \-\-sort\-time                    sort according to time string
\&
\&    Other Flags:
\&      \-r  \-\-read\-file <file>             Use <file> as input instead of STDIN
\&          \-\-completion <shell>           Generate the autocompletion script for <shell>
\&      \-f, \-\-config <file>                Configuration file (default: ~/.config/tablizer/config)
\&      \-d, \-\-debug                        Enable debugging
\&      \-h, \-\-help                         help for tablizer
\&      \-m, \-\-man                          Display manual page
\&      \-V, \-\-version                      Print program version
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Many  programs generate  tabular output.   But sometimes  you need  to
post-process these tables, you may need  to remove one or more columns
or you  may want to filter  for some pattern (See  \s-1PATTERNS\s0) or you
may need the  output in another program and need  to parse it somehow.
Standard unix tools such as \fBawk\fR\|(1), \fBgrep\fR\|(1) or \fBcolumn\fR\|(1) may help, but
sometimes it's a tedious business.
.PP
Let's take  the output of  the tool  kubectl.  It contains  cells with
withespace and they do not separate columns by \s-1TAB\s0 characters. This is
not easy to process.
.PP
You can use \fBtablizer\fR to do these and more things.
.PP
\&\fBtablizer\fR  analyses the  header  fields of  a  table, registers  the
column positions of  each header field and separates  columns by those
positions.
.PP
Without any options it reads its input from \f(CW\*(C`STDIN\*(C'\fR, but you can also
specify a  file as a  parameter. If you want  to reduce the  output by
some regular expression,  just specify it as its  first parameter. You
may also  use the  \fB\-v\fR option  to exclude all  rows which  match the
pattern. Hence:
.PP
.Vb 2
\&   # read from STDIN
\&   kubectl get pods | tablizer
\&
\&   # read a file
\&   tablizer \-r filename
\&
\&   # search for pattern in a file (works like grep)
\&   tablizer regex \-r filename
\&
\&   # search for pattern in STDIN
\&   kubectl get pods | tablizer regex
.Ve
.PP
The output looks like the original  one. You can add the option \fB\-n\fR,
then every header field will have a numer associated with it, e.g.:
.PP
.Vb 1
\&   NAME(1) READY(2) STATUS(3) RESTARTS(4) AGE(5)
.Ve
.PP
These numbers denote the column and  you can use them to specify which
columns you want to have in your output (see \s-1COLUMNS\s0:
.PP
.Vb 1
\&   kubectl get pods | tablizer \-c1,3
.Ve
.PP
You can specify the numbers in any order but output will always follow
the original order.
.PP
However, you may also just use the header names instead of numbers,
eg:
.PP

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org).

## [v1.0.14](https://github.com/TLINDEN/tablizer/tree/v1.0.14) - 2023-01-23

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.13...v1.0.14)

### Fixed

- The -D parameter could not be used together with -a.

- Fixed invalid argv handling: when the user wanted to read from stdin
  but gave an argument which was  meant as a pattern, but also existed
  as a filename, then tablizer opened the file, ignored stdin.
  
- Makefile indentation

  
### Added

- added licens notes about dependencies

- using hard coded uniseq version, see actions#3396457307

- updated dependencies (go module versions)


## [v1.0.13](https://github.com/TLINDEN/tablizer/tree/v1.0.13) - 2022-11-03

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.12...v1.0.13)

### Added

- Added command line flag to generate shell completion code

- Added an animated demo gif to the README to demonstrate the tool

### Fixed

- The `-A` flag wasn't implemented (default output mode).

- Fixed building from source on systems w/o perls pod tools,
  which is not requrired anyway since I always commit the latest
  manpage.


## [v1.0.12](https://github.com/TLINDEN/tablizer/tree/v1.0.12) - 2022-10-25

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.11...v1.0.12)

### Added

- Added support to parse CSV input

- Added CSV output support

- Added support for environment variables

### Changed

- We do not use the generated help message anymore, instead we use the
  usage from the  manpage, which we have to maintain  anyway. It looks
  better and has flag groups, which  cobra is still lacking as of this
  writing.
  
-  More refactoring  and  re-organization,  runtime configuration  now
  lives in the cfg module.
  
  
### Fixed

- Fixed  [Bug #5](https://github.com/TLINDEN/tablizer/issues/5), where
  matches have not been highlighted correctly in some rare cases.



## [v1.0.11](https://github.com/TLINDEN/tablizer/tree/v1.0.11) - 2022-10-19

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.10...v1.0.11)

### Added

- Added CI job golinter to regularly check for common mistakes.

- Added YAML output mode.

- Added more unit tests, we're over 95% in the lib module.

### Changed

- do  not use any global  variables anymore, makes the  code easier to
  maintain, understand and test

- using io.Writer  in print* functions, which is easier  to test, also
  re-implemented the print tests.

- replaced go-str2duration with my own implementation `duration2int()`.



## [v1.0.10](https://github.com/TLINDEN/tablizer/tree/v1.0.10) - 2022-10-15

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.9...v1.0.10)

### Added

- Added various sort modes: sort by time, by duration, numerical (-a -t -i)

- Added possibility to modify sort order to descending (-D)

- Added support  to specify a regexp in column  selector -c, which can
  also be mixed with numerical column spec
  
- More unit tests

### Fixed

- Column specification  allowed to specify duplicate  columns like `-c
  1,2,1,2` unchecked. Now this list will be deduplicated before use.



## [v1.0.9](https://github.com/TLINDEN/tablizer/tree/v1.0.9) - 2022-10-14

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.8...v1.0.9)

### Added

- Added Changelog, Contribution guidelines and no COC.

### Changed

- some minor changes to satisfy linter.



## [v1.0.8](https://github.com/TLINDEN/tablizer/tree/v1.0.8) - 2022-10-13

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.7...v1.0.8)

### Added

- Added sort support with the new parameter -k (like sort(1)).



## [v1.0.7](https://github.com/TLINDEN/tablizer/tree/v1.0.7) - 2022-10-11

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.6...v1.0.7)

### Added

- Added pattern highlighting support.

- Added more unit tests.

### Fixed

- Fixed extended more output in combination with -c.

- Fixed issue #4, the version string was missing.



## [v1.0.6](https://github.com/TLINDEN/tablizer/tree/v1.0.6) - 2022-10-05

[Full Changelog](https://github.com/TLINDEN/tablizer/compare/v1.0.5...v1.0.6)

### Added

- Added documentation about regexp syntax in the manpage.

- Added more unit tests.

### Changed

- Rewrote the input parser.

- Some more refactoring work has been done.



## [v1.0.5](https://github.com/TLINDEN/tablizer/tree/v1.0.5) - 2022-10-05


cmd/tablizer.go
package cmd

var manpage = `
NAME
    tablizer - Manipulate tabular output of other programs

SYNOPSIS
        Usage:
          tablizer [regex,...] [-r file] [flags]
    
        Operational Flags:
          -c, --columns string               Only show the speficied columns (separated by ,)
          -v, --invert-match                 select non-matching rows
          -n, --numbering                    Enable header numbering
          -N, --no-color                     Disable pattern highlighting
          -H, --no-headers                   Disable headers display
          -s, --separator <string>           Custom field separator (maybe char, string or :class:)
          -k, --sort-by <int|name>           Sort by column (default: 1)
          -z, --fuzzy                        Use fuzzy search [experimental]
          -F, --filter <field[!]=reg>        Filter given field with regex, can be used multiple times
          -T, --transpose-columns string     Transpose the speficied columns (separated by ,)
          -R, --regex-transposer </from/to/> Apply /search/replace/ regexp to fields given in -T
          -j, --json                         Read JSON input (must be array of hashes)
          -I, --interactive                  Interactively filter and select rows
          -g, --auto-headers                 Generate headers if there are none present in input
          -x, --custom-headers a,b,...       Use custom headers, separated by comma

        Output Flags (mutually exclusive):
          -X, --extended                     Enable extended output
          -M, --markdown                     Enable markdown table output
          -O, --orgtbl                       Enable org-mode table output
          -S, --shell                        Enable shell evaluable output
          -Y, --yaml                         Enable yaml output
          -J, --jsonout                      Enable JSON output
          -C, --csv                          Enable CSV output
          -A, --ascii                        Default output mode, ascii tabular
          -L, --hightlight-lines             Use alternating background colors for tables
          -o, --ofs <char>                   Output field separator, used by -A and -C. 
          -y, --yank-columns                 Yank specified columns (separated by ,) to clipboard,
                                             space separated

        Sort Mode Flags (mutually exclusive):
          -a, --sort-age                     sort according to age (duration) string
          -D, --sort-desc                    Sort in descending order (default: ascending)
          -i, --sort-numeric                 sort according to string numerical value
          -t, --sort-time                    sort according to time string

        Other Flags:
          -r  --read-file <file>             Use <file> as input instead of STDIN
              --completion <shell>           Generate the autocompletion script for <shell>
          -f, --config <file>                Configuration file (default: ~/.config/tablizer/config)
          -d, --debug                        Enable debugging
          -h, --help                         help for tablizer
          -m, --man                          Display manual page
          -V, --version                      Print program version

DESCRIPTION
    Many programs generate tabular output. But sometimes you need to
    post-process these tables, you may need to remove one or more columns or
    you may want to filter for some pattern (See PATTERNS) or you may need
    the output in another program and need to parse it somehow. Standard
    unix tools such as awk(1), grep(1) or column(1) may help, but sometimes
    it's a tedious business.

    Let's take the output of the tool kubectl. It contains cells with
    withespace and they do not separate columns by TAB characters. This is
    not easy to process.

    You can use tablizer to do these and more things.

    tablizer analyses the header fields of a table, registers the column
    positions of each header field and separates columns by those positions.

    Without any options it reads its input from "STDIN", but you can also
    specify a file as a parameter. If you want to reduce the output by some
    regular expression, just specify it as its first parameter. You may also
    use the -v option to exclude all rows which match the pattern. Hence:

       # read from STDIN
       kubectl get pods | tablizer

       # read a file
       tablizer -r filename

       # search for pattern in a file (works like grep)
       tablizer regex -r filename

       # search for pattern in STDIN
       kubectl get pods | tablizer regex

    The output looks like the original one. You can add the option -n, then
    every header field will have a numer associated with it, e.g.:

       NAME(1) READY(2) STATUS(3) RESTARTS(4) AGE(5)

    These numbers denote the column and you can use them to specify which
    columns you want to have in your output (see COLUMNS:

       kubectl get pods | tablizer -c1,3

    You can specify the numbers in any order but output will always follow
    the original order.

    However, you may also just use the header names instead of numbers, eg:

       kubectl get pods | tablizer -cname,status

    You can also use regular expressions with -c, eg:

       kubectl get pods | tablizer -c '[ae]'

    By default tablizer shows a header containing the names of each column.
    This can be disabled using the -H option. Be aware that this only
    affects tabular output modes. Shell, Extended, Yaml and CSV output modes
    always use the column names.

    By default, if a pattern has been speficied, matches will be
    highlighted. You can disable this behavior with the -N option.

    Use the -k option to specify by which column to sort the tabular data
    (as in GNU sort(1)). The default sort column is the first one. You can
    specify column numbers or names. Column numbers start with 1, names are
    case insensitive. You can specify multiple columns separated by comma to
    sort, but the type must be the same. For example if you want to sort
    numerically, all columns must be numbers. If you use column numbers,
    then be aware, that these are the numbers before column extraction. For
    example if you have a table with 4 columns and specify "-c4", then only
    1 column (the fourth) will be printed, however if you want to sort by
    this column, you'll have to specify "-k4".

    The default sort order is ascending. You can change this to descending
    order using the option -D. The default sort order is by alphanumeric
    string, but there are other sort modes:

    -a --sort-age
        Sorts duration strings like "1d4h32m51s".

    -i --sort-numeric
        Sorts numeric fields.

    -t --sort-time
        Sorts timestamps.

    Finally the -d option enables debugging output which is mostly useful
    for the developer.

  SEPARATOR
    The option -s can be a single character, in which case the CSV parser
    will be invoked. You can also specify a string as separator. The string
    will be interpreted as literal string unless it is a valid go regular
    expression. For example:

        -s '\t{2,}\'

    is being used as a regexp and will match two or more consecutive tabs.

        -s 'foo'

    on the other hand is no regular expression and will be used literally.

    To make live easier, there are a couple of predefined regular
    expressions, which you can specify as classes:

        * :tab:

        Matches a tab and eats spaces around it.

        * :spaces:

        Matches 2 or more spaces.

        * :pipe:

        Matches a pipe character and eats spaces around it.

        * :default:

        Matches 2 or more spaces or tab. This is the default separator if
        none is specified.

        * :nonword:

        Matches a non-word character.

        * :nondigit:

        Matches a non-digit character.

        * :special:

        Matches one or more special chars like brackets, dollar sign,
        slashes etc.

        * :nonprint:

        Matches one or more non-printable characters.

  PATTERNS AND FILTERING
    You can reduce the rows being displayed by using one or more regular
    expression patterns. The regexp language being used is the one of
    GOLANG, refer to the syntax cheat sheet here:
    <https://pkg.go.dev/regexp/syntax>.

    If you want to read a more comprehensive documentation about the topic
    and have perl installed you can read it with:

        perldoc perlre

    Or read it online: <https://perldoc.perl.org/perlre>. But please note
    that the GO regexp engine does NOT support all perl regex terms,
    especially look-ahead and look-behind.

    If you want to supply flags to a regex, then surround it with slashes
    and append the flag. The following flags are supported:

        i => case insensitive
        ! => negative match

    Example for a case insensitive search:

        kubectl get pods -A | tablizer "/account/i"

    If you use the "!" flag, then the regex match will be negated, that is,
    if a line in the input matches the given regex, but "!" is supplied,
    tablizer will NOT include it in the output.

    For example, here we want to get all lines matching "foo" but not "bar":

        cat table | tablizer foo '/bar/!'

    This would match a line "foo zorro" but not "foo bar".

    The flags can also be combined.

    You can also use the experimental fuzzy search feature by providing the
    option -z, in which case the pattern is regarded as a fuzzy search term,
    not a regexp.

    Sometimes you want to filter by one or more columns. You can do that
    using the -F option. The option can be specified multiple times and has
    the following format:

        fieldname=regexp

    Fieldnames (== columns headers) are case insensitive.

    If you specify more than one filter, both filters have to match (AND
    operation).

    These field filters can also be negated:

        fieldname!=regexp

    If the option -v is specified, the filtering is inverted.

  INTERACTIVE FILTERING
    You can also use the interactive mode, enabled with "-I" to filter and
    select rows. This mode is complementary, that is, other filter options
    are still being respected.

    To enter e filter, hit "/", enter a filter string and finish with
    "ENTER". Use "SPACE" to select/deselect rows, use "a" to select all
    (visible) rows.

    Commit your selection with "q". The selected rows are being fed to the
    requested output mode as usual. Abort with "CTRL-c", in which case the
    results of the interactive mode are being ignored and all rows are being
    fed to output.

  COLUMNS
    The parameter -c can be used to specify, which columns to display. By
    default tablizer numerizes the header names and these numbers can be
    used to specify which header to display, see example above.

    However, beside numbers, you can also use regular expressions with -c,
    also separated by comma. And you can mix column numbers with regexps.

    Lets take this table:

            PID TTY          TIME CMD
          14001 pts/0    00:00:00 bash
          42871 pts/0    00:00:00 ps
          42872 pts/0    00:00:00 sed

    We want to see only the CMD column and use a regex for this:

        ps | tablizer -s '\s+' -c C
        CMD(4)
        bash
        ps
        tablizer
        sed

    where "C" is our regexp which matches CMD.

    If a column specifier doesn't look like a regular expression, matching
    against header fields will be case insensitive. So, if you have a field
    with the name "ID" then these will all match: "-c id", "-c Id". The same
    rule applies to the options "-T" and "-F".

  TRANSPOSE FIELDS USING REGEXPS
    You can manipulate field contents using regular expressions. You have to
    tell tablizer which field[s] to operate on using the option "-T" and the
    search/replace pattern using "-R". The number of columns and patterns
    must match.

    A search/replace pattern consists of the following elements:

        /search-regexp/replace-string/

    The separator can be any valid character. Especially if you want to use
    a regexp containing the "/" character, eg:

        |search-regexp|replace-string|

    Example:

        cat t/testtable2
        NAME  DURATION
        x     10
        a     100
        z     0
        u     4
        k     6
    
        cat t/testtable2 | tablizer -T2 -R '/^\d/4/' -n
        NAME    DURATION 
        x       40      
        a       400     
        z       4       
        u       4       
        k       4

  OUTPUT MODES
    There might be cases when the tabular output of a program is way too
    large for your current terminal but you still need to see every column.
    In such cases the -o extended or -X option can be useful which enables
    *extended mode*. In this mode, each row will be printed vertically,
    header left, value right, aligned by the field widths. Here's an
    example:

        kubectl get pods | ./tablizer -o extended
            NAME: repldepl-7bcd8d5b64-7zq4l  
           READY: 1/1    
          STATUS: Running  
        RESTARTS: 1 (71m ago)  
             AGE: 5h28m

    You can of course still use a regex to reduce the number of rows
    displayed.

    The option -o shell can be used if the output has to be processed by the
    shell, it prints variable assignments for each cell, one line per row:

        kubectl get pods | ./tablizer -o extended ./tablizer -o shell
        NAME="repldepl-7bcd8d5b64-7zq4l" READY="1/1" STATUS="Running" RESTARTS="9 (47m ago)" AGE="4d23h" 
        NAME="repldepl-7bcd8d5b64-m48n8" READY="1/1" STATUS="Running" RESTARTS="9 (47m ago)" AGE="4d23h" 
        NAME="repldepl-7bcd8d5b64-q2bf4" READY="1/1" STATUS="Running" RESTARTS="9 (47m ago)" AGE="4d23h"

    You can use this in an eval loop.

    Beside normal ascii mode (the default) and extended mode there are more
    output modes available: orgtbl which prints an Emacs org-mode table and
    markdown which prints a Markdown table, yaml, which prints yaml encoding
    and CSV mode, which prints a comma separated value file.

  PUT FIELDS TO CLIPBOARD
    You can let tablizer put fields to the clipboard using the option "-y".
    This best fits the use-case when the result of your filtering yields
    just one row. For example:

        cloudctl cluster ls | tablizer -yid matchbox

    If "matchbox" matches one cluster, you can immediately use the id of
    that cluster somewhere else and paste it. Of course, if there are
    multiple matches, then all id's will be put into the clipboard separated
    by one space.

  ENVIRONMENT VARIABLES
    tablizer supports certain environment variables which use can use to
    influence program behavior. Commandline flags have always precedence
    over environment variables.

    <T_HEADER_NUMBERING> - enable numbering of header fields, like -n.
    <T_COLUMNS> - comma separated list of columns to output, like -c
    <NO_COLORS> - disable colorization of matches, like -N

  COMPLETION
    Shell completion for command line options can be enabled by using the
    --completion flag. The required parameter is the name of your shell.
    Currently supported are: bash, zsh, fish and powershell.

    Detailed instructions:

    Bash:
           source <(tablizer --completion bash)

        To load completions for each session, execute once:

          # Linux:
          $ tablizer --completion bash > /etc/bash_completion.d/tablizer

          # macOS:
          $ tablizer --completion bash > $(brew --prefix)/etc/bash_completion.d/tablizer

    Zsh:
        If shell completion is not already enabled in your environment, you
        will need to enable it. You can execute the following once:

          echo "autoload -U compinit; compinit" >> ~/.zshrc

        To load completions for each session, execute once:

          $ tablizer --completion zsh > "${fpath[1]}/_tablizer"

        You will need to start a new shell for this setup to take effect.

    fish:
           tablizer --completion fish | source

        To load completions for each session, execute once:

           tablizer --completion fish > ~/.config/fish/completions/tablizer.fish

    PowerShell:
           tablizer --completion powershell | Out-String | Invoke-Expression

        To load completions for every new session, run:

           tablizer --completion powershell > tablizer.ps1

        and source this file from your PowerShell profile.

CONFIGURATION AND COLORS
    YOu can put certain configuration values into a configuration file in
    HCL format. By default tablizer looks for
    "$HOME/.config/tablizer/config", but you can provide one using the
    parameter "-f".

    In the configuration the following variables can be defined:

        BG             = "lightGreen"
        FG             = "white"
        HighlightBG    = "lightGreen"
        HighlightFG    = "white"
        NoHighlightBG  = "white"
        NoHighlightFG  = "lightGreen"
        HighlightHdrBG = "red"
        HighlightHdrFG = "white"

    The following color definitions are available:

    black, blue, cyan, darkGray, default, green, lightBlue, lightCyan,
    lightGreen, lightMagenta, lightRed, lightWhite, lightYellow, magenta,
    red, white, yellow

    The Variables FG and BG are being used to highlight matches. The other
    *FG and *BG variables are for colored table output (enabled with the
    "-L" parameter).

    Colorization can be turned off completely either by setting the
    parameter "-N" or the environment variable NO_COLOR to a true value.

BUGS
    In order to report a bug, unexpected behavior, feature requests or to
    submit a patch, please open an issue on github:
    <https://github.com/TLINDEN/tablizer/issues>.

LICENSE
    This software is licensed under the GNU GENERAL PUBLIC LICENSE version
    3.

    Copyright (c) 2022-2024 by Thomas von Dein

    This software uses the following GO modules:

    repr (https://github.com/alecthomas/repr)
        Released under the MIT License, Copyright (c) 2016 Alec Thomas

    cobra (https://github.com/spf13/cobra)
        Released under the Apache 2.0 license, Copyright 2013-2022 The Cobra
        Authors

    dateparse (github.com/araddon/dateparse)
        Released under the MIT License, Copyright (c) 2015-2017 Aaron Raddon

    color (github.com/gookit/color)
        Released under the MIT License, Copyright (c) 2016 inhere

    tablewriter (github.com/olekukonko/tablewriter)
        Released under the MIT License, Copyright (c) 201 by Oleku Konko

    yaml (gopkg.in/yaml.v3)
        Released under the MIT License, Copyright (c) 2006-2011 Kirill
        Simonov

    bubble-table (https://github.com/Evertras/bubble-table)
        Released under the MIT License, Copyright (c) 2022 Brandon Fulljames

AUTHORS
    Thomas von Dein tom AT vondein DOT org

`
var usage = `

Usage:
  tablizer [regex,...] [-r file] [flags]

Operational Flags:
  -c, --columns string               Only show the speficied columns (separated by ,)
  -v, --invert-match                 select non-matching rows
  -n, --numbering                    Enable header numbering
  -N, --no-color                     Disable pattern highlighting
  -H, --no-headers                   Disable headers display
  -s, --separator <string>           Custom field separator (maybe char, string or :class:)
  -k, --sort-by <int|name>           Sort by column (default: 1)
  -z, --fuzzy                        Use fuzzy search [experimental]
  -F, --filter <field[!]=reg>        Filter given field with regex, can be used multiple times
  -T, --transpose-columns string     Transpose the speficied columns (separated by ,)
  -R, --regex-transposer </from/to/> Apply /search/replace/ regexp to fields given in -T
  -j, --json                         Read JSON input (must be array of hashes)
  -I, --interactive                  Interactively filter and select rows
  -g, --auto-headers                 Generate headers if there are none present in input
  -x, --custom-headers a,b,...       Use custom headers, separated by comma

Output Flags (mutually exclusive):
  -X, --extended                     Enable extended output
  -M, --markdown                     Enable markdown table output
  -O, --orgtbl                       Enable org-mode table output
  -S, --shell                        Enable shell evaluable output
  -Y, --yaml                         Enable yaml output
  -J, --jsonout                      Enable JSON output
  -C, --csv                          Enable CSV output
  -A, --ascii                        Default output mode, ascii tabular
  -L, --hightlight-lines             Use alternating background colors for tables
  -o, --ofs <char>                   Output field separator, used by -A and -C. 
  -y, --yank-columns                 Yank specified columns (separated by ,) to clipboard,
                                     space separated

Sort Mode Flags (mutually exclusive):
  -a, --sort-age                     sort according to age (duration) string
  -D, --sort-desc                    Sort in descending order (default: ascending)
  -i, --sort-numeric                 sort according to string numerical value
  -t, --sort-time                    sort according to time string

Other Flags:
  -r  --read-file <file>             Use <file> as input instead of STDIN
      --completion <shell>           Generate the autocompletion script for <shell>
  -f, --config <file>                Configuration file (default: ~/.config/tablizer/config)
  -d, --debug                        Enable debugging
  -h, --help                         help for tablizer
  -m, --man                          Display manual page
  -V, --version                      Print program version


`
